FIX make creation of dataset deterministic in SGD

PierreAttard opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
Fixes #19603 


#### What does this implement/fix? Explain your changes.
`check_random_state` moved up before `make_dataset` in order to use the same random state during the whole process. 
This change occurs in classes `SGDRegressor, SparseSGDRegressor`, `fit` method. 

#### Any other comments?

I had to adapt the existing unitest `test_validation_set_not_used_for_training` in file `test_sgd.py` : 

With this change, the random state is "used" one time before the following line :
```python
validation_mask = self._make_validation_split(y)
```
So, I reproduce this effect in the unitest with those changes : 
```python
        rng = np.random.RandomState(seed)
        rng.randint(1, np.iinfo(np.int32).max)
        cv = ShuffleSplit(test_size=validation_fraction,
                          random_state=rng)
```

I am not totally satisfied, what do you think about ?

If you're OK, I add text in **what news** doc and add a non-regression test to check that the model is fully deterministic when seeding the random state like said @ogrisel 

Thanks in advance for responses !

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

PierreAttard committed:
check_random_state moved up in order to use it in make_dataset and unitest adaptation

ogrisel reviewed (commented):
Could you please add a non-regression test that check that calling `SGDRegressor` twice on the same data without waiting for convergence (for instance using `max_iter=1`, gives the same `model.coef_` and `model.intercept_` attribute?

ogrisel commented:
You will also need to document the fix in `doc/whats_new/v0.24.rst` for the 0.24.2 release.

ogrisel commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -387,8 +387,10 @@ def test_validation_set_not_used_for_training(klass):
+        rng.randint(1, np.iinfo(np.int32).max)
         cv = ShuffleSplit(test_size=validation_fraction,
-                          random_state=seed)
+                          random_state=rng)
```
Sorry I had not read the description of the PR...

I think we need to rework the test to make it less hackish.

ogrisel replied:
You can probably just move the creation of the validation split `validation_score_cb` just after the call to `check_random_state`.

It's still dependent a bit on internal details but the test would appear less convoluted.

PierreAttard replied:
That's what I explain in the PR message. 


> #### Any other comments?
> 
> I had to adapt the existing unitest `test_validation_set_not_used_for_training` in file `test_sgd.py` :
> 
> With this change, the random state is "used" one time before the following line :
> 
> ```python
> validation_mask = self._make_validation_split(y)
> ```
> 
> So, I reproduce this effect in the unitest with those changes :
> 
> ```python
>         rng = np.random.RandomState(seed)
>         rng.randint(1, np.iinfo(np.int32).max)
>         cv = ShuffleSplit(test_size=validation_fraction,
>                           random_state=rng)
> ```
> 
> I am not totally satisfied, what do you think about ?
> 
> If you're OK, I add text in **what news** doc and add a non-regression test to check that the model is fully deterministic when seeding the random state like said @ogrisel
> 
> Thanks in advance for responses !



If you use the seed, the results will be different because inside `_fit_regressor` method, the random object is **run** a first time with the given `seed` before the use of `_make_validation_split`. That is why in the test, in order to reproduce that behavior, I create `rng` which is **run** a first time before the validaiton split. But I am not really convinced. 

Sorry if I am not clear !

PierreAttard replied:
The **run** term is probably wrong but means in this case : 

```python
rng.randint(1, np.iinfo(np.int32).max)
```

ogrisel replied:
Alright I get it now...

Still I don't like the way this test is written. I think we should rewrite it completely. For instance by fitting a model on a random target, with and without early stopping on a validation split. The accuracy measured on `(X_train, y_train)` and the value of the `n_iter_` attribute of the model without validation set early stopping should be much higher than the model with early stopping.

jeremiedbb replied:
I made it such that there's no change in the model. I explained below that SGD regressors have no use of the seed attribute of the generated dataset. So I think it's better to just document that in the PR and leave this test as is. In addition the test you propose is not guaranteed, but only quite probable.

ogrisel commented:
@PierreAttard gentle reminder that there are still some comments to address in this PR ;)

PierreAttard commented:
Yes, indeed ! I'll try to continue the work next week !

glemaitre committed:
address ogrisel comments

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/PierreAttard/19716

glemaitre committed:
DOC add whats new

glemaitre commented:
I replaced the test as @ogrisel mentioned and added an entry in what's new.

@ogrisel can you have a quick look to see if the test reflect what you had in mind.

PierreAttard commented:
Thanks a lot @glemaitre . I didn't have much time. I check as soon as possible.

glemaitre commented:
I see that the CIs are failing. It should be linked to the fact that we changed the way the random state is working.
This is not a big deal, we need to mention it in the what's new but I am thinking that it could be better to move this PR to be included in 1.0 instead of 0.24.2 because it could be weird to have a change of behaviour in a bug fix release.

@ogrisel WDYT?

PierreAttard commented:
Hi @glemaitre, I launched the test ```test_validation_set_not_used_for_training``` several times. 
Most of time, the 4 tests pass but some time, one of them does not pass, and only for ```SGDClassifier``` estimator. 

Did you not that too ?
Shouldn't we make it reproducible ?

Sorry again for the delay of the review.

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/PierreAttard/19716

glemaitre committed:
TST make data generation deterministic

glemaitre commented:
I solved the conflict in this PR and make the test deterministic by creating a deterministic dataset indeed.

glemaitre reviewed (dismissed):
LGTM

glemaitre committed:
docstring

PierreAttard committed:
Merge branch 'scikit-learn:main' into random-state-not-used-in-make_dataset

adrinjalali reviewed (commented):
otherwise LGTM

adrinjalali commented:
Please merge with `main` and fix the merge conflict as well :)

Commit:
Merge branch 'main' into random-state-not-used-in-make_dataset

# Conflicts:
#	doc/whats_new/v1.0.rst

Commit:
revert merge with main

Commit:
merge changelog file

Commit:
revert the revert merge

Commit:
merge chnagelog file

Commit:
merge chnagelog file V2 remove PR 19716

Commit:
merge chnagelog file V2 add PR 19716

PierreAttard committed:
Merge branch 'main' into random-state-not-used-in-make_dataset

PierreAttard commented:
Hi, sorry for the last useless commit (I didn't see the v1.0.rst conflict).

Commit:
remove ".git-blame-ignore-revs" file in order to match with sdist

adrinjalali commented:
@PierreAttard I also left some comments, did you see them?

Commit:
Merge branch 'main' into random-state-not-used-in-make_dataset

Commit:
Add digit in doc for class PassiveAggressiveRegressor.

Wrongly name variable replaced with a right name (class-name-like replaced by `klass`)

PierreAttard commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -444,47 +443,44 @@ def test_adaptive_longer_than_constant(klass):
-    shuffle = False
-    max_iter = 10
-    clf1 = klass(
+def test_validation_set_not_used_for_training(SGDEstimator, X, y):
```
Yes, it's more readable.

PierreAttard commented on sklearn/linear_model/_passive_aggressive.py:
```diff
@@ -416,11 +416,11 @@ class PassiveAggressiveRegressor(BaseSGDRegressor):
     PassiveAggressiveRegressor(max_iter=100, random_state=0)
     >>> print(regr.coef_)
-    [20.48736655 34.18818427 67.59122734 87.94731329]
+    [20.5... 34.2... 67.6... 87.9...]
```
Yes.

PierreAttard commented:
@adrinjalali No I didn't till your comment ! I took into account your comments in the last commits.

adrinjalali commented on .git-blame-ignore-revs:
```diff
@@ -1,18 +0,0 @@
-# Since git version 2.23, git-blame has a feature to ignore
```
you probably deleted this file by mistake

PierreAttard replied:
In fact, I had some failed messages because of that file, so I intentionally removed it, but was a mistake : 

![image](https://user-images.githubusercontent.com/11873372/132005457-618e2e5b-85d6-43f2-9461-b18adcb01ab5.png)

Commit:
add '.git-blame-ignore-revs' file

Commit:
merge changelog file V2 add PR 19716

PierreAttard committed:
Merge branch 'main' into random-state-not-used-in-make_dataset

jjerphan reviewed (dismissed):
Thank you for working on this fix, @PierreAttard.

Here are some suggestions, a few of which might be corrected when merging `main` in this PR branch.

jjerphan commented:
Hello @PierreAttard,

Do you still have some time to work on this PR?

jeremiedbb committed:
Merge branch 'master' into pr/PierreAttard/19716

jeremiedbb committed:
merge clean-up

jeremiedbb committed:
more cln

jeremiedbb committed:
better document how random_state is used in dataset

Signed-off-by: jeremie du boisberranger <jeremiedbb@yahoo.fr>

jeremiedbb committed:
no model is changed

Signed-off-by: jeremie du boisberranger <jeremiedbb@yahoo.fr>

jeremiedbb commented on sklearn/linear_model/_passive_aggressive.py:
```diff
@@ -428,11 +428,11 @@ class PassiveAggressiveRegressor(BaseSGDRegressor):
+    [-0.0055367]
     >>> print(regr.predict([[0, 0, 0, 0]]))
-    [-0.02306214]
+    [-0.0055367]
```
made it such that no model is changed

ogrisel committed:
TST add non regression test about random seed control for SGD models

jeremiedbb commented on sklearn/linear_model/tests/test_sgd.py:
```diff
@@ -2140,3 +2140,41 @@ def test_loss_squared_loss_deprecated(Estimator):
+)
+def test_sgd_random_state(Estimator, global_random_seed):
+    # Train the same model on the same data without converging and check that we
+    # can reprodicible results by fixing the random seed.
```
```suggestion
    # get reproducible results by fixing the random seed.
```

jeremiedbb committed:
Update sklearn/linear_model/tests/test_sgd.py

ogrisel reviewed (approved):
I closed #19603 because there is no bug in the end but +1 for merging this PR to make the code less surprising.

ogrisel commented:
Note, I have checked that the new test passes with `SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all"`.

jjerphan reviewed (approved):
LGTM. Thank you @PierreAttard and @jeremiedbb.

jeremiedbb commented:
The failure should be unrelated to the PR (see https://github.com/scikit-learn/scikit-learn/issues/23014) but it's not triggered in other PRs. I'm starting it again to see if it's deterministic

jeremiedbb commented:
That random failure was really weird...
Let's still merge this PR and try to find what's going on in #23014

jeremiedbb commented:
Thanks @PierreAttard !
